src/main/java/ca/corefacility/bioinformatics/irida/service/impl/sample/SampleServiceImpl.java

```diff
@@ -172,6 +180,76 @@ public Sample update(Sample object) {
 		return super.update(object);
 	}
 
+	/**
+	 * {@inheritDoc}
+	 */
+	@PreAuthorize("hasPermission(#sample, 'canReadSample')")
+	@Override
+	public Set<MetadataEntry> getMetadataForSample(Sample sample) {
+		return metadataEntryRepository.getMetadataForSample(sample);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@PreAuthorize("hasPermission(#s, 'canUpdateSample')")
+	@Transactional
+	public Sample updateSampleMetadata(Sample s, Set<MetadataEntry> metadataToSet) {
+		Set<MetadataEntry> currentMetadata = getMetadataForSample(s);
+
+		metadataEntryRepository.deleteAll(currentMetadata);
+
+		for(MetadataEntry e : metadataToSet){
+			e.setSample(s);
+		}
+
+		metadataEntryRepository.saveAll(metadataToSet);
+
+		s = read(s.getId());
+
+		return s;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@PreAuthorize("hasPermission(#s, 'canUpdateSample')")
+	@Transactional
+	public Sample mergeSampleMetadata(Sample s, Set<MetadataEntry> metadataToAdd) {
+		Set<MetadataEntry> currentMetadata = getMetadataForSample(s);
+
+		// loop through entry set and see if it already exists
+		for (MetadataEntry newMetadataEntry : metadataToAdd) {
+			MetadataTemplateField field = newMetadataEntry.getField();
+			newMetadataEntry.setSample(s);
+
+			Optional<MetadataEntry> metadataEntryForField = currentMetadata.stream()
+					.filter(e -> e.getField()
+							.equals(field))
+					.findFirst();
```

Comment:
I suspect the `findFirst()` here causes issues where we can have many duplicated entries in our database for metadata fields. See comment in general comments.

Reply:
Added a fix in 000defa30f which deletes the old metadata entry if the types differ.  We're seeing this sort of thing in a bunch of places with JPA that even if we save a new collection without the old entity, it sometimes stays connected.  The result has generally been to actually delete the old entity before saving the new collection.  This appears to resolve the above issue number 2.